# Übungszettel 7: Entscheidungsbäume

## Maschinelles Lernen - WiSe 23/24

### Abgabe 13.12.2023, 23:55 Uhr

*Hinweise:*
- Übungsaufgaben **müssen** in Gruppen von 3-4 Personen abgegeben werden. **Einzelabgaben werden nicht korrigiert bzw. bewertet.**
- Es wird pro Übungszettel nur eine Aufgabe bewertet, die übrigen Aufgaben dienen zur selbstständigen Vertiefung des Vorlesungsstoffs. Für diese Aufgaben werden nach der Abgabe Musterlösungen bereitgestellt.
- Die Lösungen sollen in diesem IPython Notebook realisiert werden, wobei Teilaufgaben und Zwischenergebnisse ausgegeben bzw. visualisiert werden sollen.
- Für die Abgabe sollen Sie dieses IPython Notebook und ggf. zugehörige Dateien in ein **Ziparchiv** packen und im Ilias hochladen. Das Ziparchiv soll nach folgendem Muster benannt werden:
`UebungXX_Nachname1_Nachname2_Nachname3.zip`, wobei die Nachnamen in alphabetischer Reihenfolge angegeben und Umlaute ggf. ersetzt werden sollen. Bei Nichtbefolgung dieser Vorgabe können Punkte abgezogen werden.

---
## **Aufgabe 1:  Regressionsbäume (bewertet: 3+1+1 Punkte)**

Gegeben sei folgende Beispielmenge:

| id | A1 | A2 | A3 | A4 | Value |
| -- | -- | -- | -- | -- | ----- |
| 0  | C  | K  | T  | X  | 0.28  |
| 1  | B  | J  | S  | X  | 0.50  |
| 2  | C  | J  | S  | Z  | 0.35  |
| 3  | B  | I  | R  | Y  | 5.50  |
| 4  | A  | J  | T  | Z  | 0.35  |
| 5  | A  | K  | S  | Z  | 0.80  |
| 6  | C  | I  | R  | Y  | 5.10  |
| 7  | A  | I  | R  | Y  | 5.70  |
| 8 | C  | I  | S  | Y  | 0.76  |
| 9 | B  | I  | S  | X  | 1.03  |
| 10 | B  | K  | R  | Y  | 0.46  |
| 11 | C  | K  | T  | Z  | 0.39  |
| 12 | B  | K  | S  | X  | 0.28  |
| 13 | A  | K  | R  | X  | 1.10  |

a) Erzeugen Sie einen Regressionsbaum mittels des wie folgt modifizierten Verfahrens ID3. 
Verwenden Sie das Maß Standard Deviation Reduction (SDR) zur Auswahl der Tests und den Mittelwert der Instanzen eines Blattes als Vorhersagewert. Dabei  berechnet sich die Standardabweichung einer Menge von Werten $x_1,···, x_k$ mit dem Mittelwert $\bar{x}$ sich wie folgt:

$$ SD(x_1,···, x_k) = \sqrt{\frac{1}{k-1} \sum_{i=1}^k (x_i -\bar{x})^2} $$

und Standard Deviation Reduction SDR wie folgt:

$$ SDR(S,A) = SD(S) − \sum_{i} \frac{|S_i|}{|S|} SD(S_i)$$ 

Hierbei soll ein Knoten, sobald er weniger als 3 Instanzen abdeckt, nicht weiter aufgeteilt und zu einem Blatt umgewandelt werden. Sollte bei einem Test ein Testausgang keine Instanzen abdecken, fließt er nicht in die Berechnung des SDRs ein und soll, da keine Daten für ihn vorhanden sind, als Blatt verwendet werden, das den Mittelwert seines Elternknotens vorhersagt. Im Falle zweier gleichwertiger Tests überlegen Sie sich, wie man diesen Konflikt lösen kann.

In [35]:
import numpy as np

# Daten
A1 = np.array(['C','B','C','B','A','A','C','A','C','B','B','C','B','A'])
A2 = np.array(['K', 'J', 'J', 'I', 'J', 'K', 'I', 'I', 'I', 'I', 'K', 'K', 'K', 'K'])
A3 = np.array(['T', 'S', 'S', 'R', 'T', 'S', 'R', 'R', 'S', 'S', 'R', 'T', 'S', 'R'])
A4 = np.array(['X', 'X', 'Z', 'Y', 'Z', 'Z', 'Y', 'Y', 'Y', 'X', 'Y', 'Z', 'X', 'X'])
values = np.array([0.28, 0.50, 0.35, 5.50, 0.35, 0.80, 5.10, 5.70, 0.76, 1.03, 0.46, 0.39, 0.28, 1.10])



In [22]:
#Initialisierung
#Wurzel des Baumes, die alle verfügbaren Instanzen (Datenpunkte) enthält.
# Berechnung der Gesamtstandardabweichung

#Values
mean_value = np.mean(values)
print('mean_value', mean_value)
sd_total = np.sqrt(np.sum((values - mean_value)**2) / (len(values) - 1)) #Berechnung der Standardabweichung SD
print('sd_total', sd_total)



mean_value 1.614285714285715
sd_total 2.0897394548151187


In [23]:
#Dataset A1
#For A
#calculate the mean value of A2 where A2 = A
mean_A = np.mean(values[A1 == 'A'])
print('mean_A', mean_A)
sd_A = np.sqrt(np.sum((values[A1 == 'A'] - mean_A)**2) / (len(values[A1 == 'A']) - 1)) #Berechnung der Standardabweichung SD
print('sd_A', sd_A)
#For B
#calculate the mean value of A2 where A2 = B
mean_B = np.mean(values[A1 == 'B'])
print('mean_B', mean_B)
sd_B = np.sqrt(np.sum((values[A1 == 'B'] - mean_B)**2) / (len(values[A1 == 'B']) - 1)) #Berechnung der Standardabweichung SD
print('sd_B', sd_B)
#For C
#calculate the mean value of A2 where A2 = C
mean_C = np.mean(values[A1 == 'C'])
print('mean_C', mean_C)
sd_C = np.sqrt(np.sum((values[A1 == 'C'] - mean_C)**2) / (len(values[A1 == 'C']) - 1)) #Berechnung der Standardabweichung SD
print('sd_C', sd_C)

#standard deviation reduction
sdr_A1 = sd_total - np.sum ( len(values[A1 == 'A'])/len(values) * sd_A + len(values[A1 == 'B'])/len(values) * sd_B + len(values[A1 == 'C'])/len(values) * sd_C)
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A1', sdr_A1)

mean_A 1.9874999999999998
sd_A 2.494118080604846
mean_B 1.554
sd_B 2.223528727046269
mean_C 1.376
sd_C 2.090078945877404
sdr_A1 -0.1634398799732919


In [24]:
#Dataset A2
#For A
#calculate the mean value of A2 where A2 = K
mean_K = np.mean(values[A2 == 'K'])
print('mean_K', mean_K)
sd_K = np.sqrt(np.sum((values[A2 == 'K'] - mean_K)**2) / (len(values[A2 == 'K']) - 1)) #Berechnung der Standardabweichung SD
print('sd_K', sd_K)
#For J
#calculate the mean value of A2 where A2 = J
mean_J = np.mean(values[A2 == 'J'])
print('mean_J', mean_J)
sd_J = np.sqrt(np.sum((values[A2 == 'J'] - mean_J)**2) / (len(values[A2 == 'J']) - 1)) #Berechnung der Standardabweichung SD
print('sd_J', sd_J)
#For C
#calculate the mean value of A2 where A2 = I
mean_I = np.mean(values[A2 == 'I'])
print('mean_I', mean_I)
sd_I = np.sqrt(np.sum((values[A2 == 'I'] - mean_I)**2) / (len(values[A2 == 'I']) - 1)) #Berechnung der Standardabweichung SD
print('sd_I', sd_I)

#standard deviation reduction
sdr_A2 = sd_total - np.sum ( len(values[A2 == 'K'])/len(values) * sd_K + len(values[A2 == 'J'])/len(values) * sd_J + len(values[A2 == 'I'])/len(values) * sd_I)
#sdr_A2 = sd_total - (len(values[A2 == 'K']) / len(values) * sd_K) - (len(values[A2 == 'J']) / len(values) * sd_B) - (len(values[A2 == 'I']) / len(values) * sd_C)
print('sdr_A2', sdr_A2)

mean_K 0.5516666666666666
sd_K 0.32999494945630103
mean_J 0.39999999999999997
sd_J 0.08660254037844388
mean_I 3.6180000000000008
sd_I 2.4969421298860732
sdr_A2 1.03799031429344


In [25]:
#Dataset A3
#For A
#calculate the mean value of A3 where A3 = T
mean_T = np.mean(values[A3 == 'T'])
print('mean_T', mean_T)
sd_T = np.sqrt(np.sum((values[A3 == 'T'] - mean_T)**2) / (len(values[A3 == 'T']) - 1)) #Berechnung der Standardabweichung SD
print('sd_T', sd_T)
#For S
#calculate the mean value of A3 where A3 = S
mean_S = np.mean(values[A3 == 'S'])
print('mean_S', mean_S)
sd_S = np.sqrt(np.sum((values[A3 == 'S'] - mean_S)**2) / (len(values[A3 == 'S']) - 1)) #Berechnung der Standardabweichung SD
print('sd_S', sd_S)
#For C
#calculate the mean value of A3 where A3 = I
mean_R = np.mean(values[A3 == 'R'])
print('mean_R', mean_R)
sd_R = np.sqrt(np.sum((values[A3 == 'R'] - mean_R)**2) / (len(values[A3 == 'R']) - 1)) #Berechnung der Standardabweichung SD
print('sd_R', sd_R)

#standard deviation reduction
sdr_A3 = sd_total - np.sum ( len(values[A3 == 'T'])/len(values) * sd_T + len(values[A3 == 'S'])/len(values) * sd_S + len(values[A3 == 'R'])/len(values) * sd_R)
#sdr_A3 = sd_total - (len(values[A3 == 'T']) / len(values) * sd_T) - (len(values[A3 == 'S']) / len(values) * sd_B) - (len(values[A3 == 'R']) / len(values) * sd_C)
print('sdr_A3', sdr_A3)

mean_T 0.34
sd_T 0.05567764362830021
mean_S 0.6200000000000001
sd_S 0.2908607914449798
mean_R 3.5720000000000005
sd_R 2.5678629246904907
sdr_A3 1.0360600046003163


In [26]:
#Dataset A4
#For A
#calculate the mean value of A4 where A4 = T
mean_X = np.mean(values[A4 == 'X'])
print('mean_X', mean_X)
sd_X = np.sqrt(np.sum((values[A4 == 'X'] - mean_X)**2) / (len(values[A4 == 'X']) - 1)) #Berechnung der Standardabweichung SD
print('sd_X', sd_X)
#For S
#calculate the mean value of A4 where A4 = S
mean_Z = np.mean(values[A4 == 'Z'])
print('mean_Z', mean_Z)
sd_Z = np.sqrt(np.sum((values[A4 == 'Z'] - mean_Z)**2) / (len(values[A4 == 'Z']) - 1)) #Berechnung der Standardabweichung SD
print('sd_Z', sd_Z)
#For C
#calculate the mean value of A4 where A4 = I
mean_Y = np.mean(values[A4 == 'Y'])
print('mean_Y', mean_Y)
sd_Y = np.sqrt(np.sum((values[A4 == 'Y'] - mean_I)**2) / (len(values[A4 == 'Y']) - 1)) #Berechnung der Standardabweichung SD
print('sd_Y', sd_Y)

#standard deviation reduction
sdr_A4 = sd_total - np.sum ( len(values[A4 == 'X'])/len(values) * sd_X + len(values[A4 == 'Z'])/len(values) * sd_Z + len(values[A4 == 'Y'])/len(values) * sd_Y)
#sdr_A4 = sd_total - (len(values[A4 == 'X']) / len(values) * sd_T) - (len(values[A4 == 'Z']) / len(values) * sd_B) - (len(values[A4 == 'Y']) / len(values) * sd_C)
print('sdr_A4', sdr_A4)

mean_X 0.638
sd_X 0.4007742506698753
mean_Z 0.47250000000000003
sd_Z 0.2191460700081113
mean_Y 3.5040000000000004
sd_Y 2.6558473224189676
sdr_A4 0.9354757301382144


In [27]:
#Choose the highest sdr (Choose the best Test)
sdr_list = [sdr_A1, sdr_A2, sdr_A3, sdr_A4]
sdr_maxValue = max(sdr_list)
print(sdr_maxValue, 'is in sdr_list with index ', sdr_list.index(sdr_maxValue) ) #sdr_2

if(sdr_maxValue == sdr_A1):
    print('A1')
elif(sdr_maxValue == sdr_A2):
    print('A2')
elif(sdr_maxValue == sdr_A3):
    print('A3')
elif(sdr_maxValue == sdr_A4):
    print('A4')

#find the largest sd in A2
sd_list = [sd_K, sd_J, sd_I]
sd_maxValue = max(sd_list)
print(sd_maxValue,' is in sd_list with index', sd_list.index(sd_maxValue)) #sd_K




1.03799031429344 is in sdr_list with index  1
A2
2.4969421298860732  is in sd_list with index 2


In [36]:
#take all the index in A2 where A2 = I
index_list = np.where(A2 == 'I')
print(index_list) # 

#take all the element in A1 with index in index_list
A1 = A1[index_list]
A3 = A3[index_list]
A4 = A4[index_list]
values = values[index_list]

#Values
mean_value = np.mean(values)
print('mean_value', mean_value)
sd_total = np.sqrt(np.sum((values - mean_value)**2) / (len(values) - 1)) #Berechnung der Standardabweichung SD
print('sd_total', sd_total)

(array([3, 6, 7, 8, 9], dtype=int64),)
mean_value 3.6180000000000008
sd_total 2.4969421298860732


In [37]:
#Dataset A1
#For A
#calculate the mean value of A2 where A2 = A
#count A1 == 'A'

if(np.sum(A1 == 'A') <= 1):
    mean_A = 0;sd_A = 0
else:    
    mean_A = np.mean(values[A1 == 'A'])
    sd_A = np.sqrt(np.sum((values[A1 == 'A'] - mean_A)**2) / (len(values[A1 == 'A']) - 1)) #Berechnung der Standardabweichung SD
print('mean_A', mean_A)
print('sd_A', sd_A)
#For B
#calculate the mean value of A2 where A2 = B
if(np.sum(A1 == 'B') <= 1):
    mean_B = 0
    sd_B = 0
else:    
    mean_B = np.mean(values[A1 == 'B'])
    sd_B = np.sqrt(np.sum((values[A1 == 'B'] - mean_B)**2) / (len(values[A1 == 'B']) - 1)) #Berechnung der Standardabweichung SD
print('mean_B', mean_B)
print('sd_B', sd_B)
#For C
#calculate the mean value of A2 where A2 = C
if(np.sum(A1 == 'C') <= 1):
    mean_C = 0
    sd_C = 0
else:    
    mean_C = np.mean(values[A1 == 'C'])
    sd_C = np.sqrt(np.sum((values[A1 == 'C'] - mean_C)**2) / (len(values[A1 == 'C']) - 1)) #Berechnung der Standardabweichung SD
print('mean_C', mean_C)
print('sd_C', sd_C)

#standard deviation reduction
sdr_A1 = sd_total - np.sum ( len(values[A1 == 'A'])/len(values) * sd_A + len(values[A1 == 'B'])/len(values) * sd_B + len(values[A1 == 'C'])/len(values) * sd_C)
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A1', sdr_A1)

print('------------')

if(np.sum(A3 == 'R') <= 1):
    mean_R = 0;sd_R = 0
else:    
    mean_R = np.mean(values[A3 == 'R'])
    sd_R = np.sqrt(np.sum((values[A3 == 'R'] - mean_R)**2) / (len(values[A3 == 'R']) - 1)) #Berechnung der Standardabweichung SD
print('mean_R', mean_R)
print('sd_R', sd_R)
#For B
#calculate the mean value of A2 where A2 = B
if(np.sum(A3 == 'S') <= 1):
    mean_S = 0
    sd_S = 0
else:    
    mean_S = np.mean(values[A3 == 'S'])
    sd_S = np.sqrt(np.sum((values[A3 == 'S'] - mean_S)**2) / (len(values[A3 == 'S']) - 1)) #Berechnung der Standardabweichung SD
print('mean_B', mean_S)
print('sd_B', sd_S)

#standard deviation reduction
sdr_A3 = sd_total - np.sum ( len(values[A3 == 'R'])/len(values) * sd_R + len(values[A3 == 'S'])/len(values) * sd_S )
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A3', sdr_A3)

print('------------')

if(np.sum(A4 == 'Y') <= 1):
    mean_Y = 0;sd_Y = 0
else:    
    mean_Y = np.mean(values[A4 == 'Y'])
    sd_Y = np.sqrt(np.sum((values[A4 == 'Y'] - mean_Y)**2) / (len(values[A4 == 'Y']) - 1)) #Berechnung der Standardabweichung SD
print('mean_Y', mean_Y)
print('sd_Y', sd_Y)
#For B
#calculate the mean value of A2 where A2 = B
if(np.sum(A4 == 'X') <= 1):
    mean_X = 0
    sd_X = 0
else:    
    mean_X = np.mean(values[A4 == 'X'])
    sd_X = np.sqrt(np.sum((values[A4 == 'X'] - mean_X)**2) / (len(values[A4 == 'X']) - 1)) #Berechnung der Standardabweichung SD
print('mean_X', mean_X)
print('sd_X', sd_X)

#standard deviation reduction
sdr_A4 = sd_total - np.sum ( len(values[A4 == 'Y'])/len(values) * sd_Y + len(values[A4 == 'X'])/len(values) * sd_X )
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A4', sdr_A4)




mean_A 0
sd_A 0
mean_B 3.265
sd_B 3.1607673119038675
mean_C 2.9299999999999997
sd_C 3.068843430349616
sdr_A1 0.0050978329846795845
------------
mean_R 5.433333333333334
sd_R 0.3055050463303896
mean_B 0.895
sd_B 0.19091883092036785
sdr_A3 2.2372715697196925
------------
mean_Y 4.265000000000001
sd_Y 2.3499432617264047
mean_X 0
sd_X 0
sdr_A4 0.6169875205049493


In [39]:
import numpy as np

# Daten
A1 = np.array(['C','B','C','B','A','A','C','A','C','B','B','C','B','A'])
A2 = np.array(['K', 'J', 'J', 'I', 'J', 'K', 'I', 'I', 'I', 'I', 'K', 'K', 'K', 'K'])
A3 = np.array(['T', 'S', 'S', 'R', 'T', 'S', 'R', 'R', 'S', 'S', 'R', 'T', 'S', 'R'])
A4 = np.array(['X', 'X', 'Z', 'Y', 'Z', 'Z', 'Y', 'Y', 'Y', 'X', 'Y', 'Z', 'X', 'X'])
values = np.array([0.28, 0.50, 0.35, 5.50, 0.35, 0.80, 5.10, 5.70, 0.76, 1.03, 0.46, 0.39, 0.28, 1.10])
#take all the index in A2 where A2 = I
index_list = np.where(A2 == 'K')
print(index_list) # 

#take all the element in A1 with index in index_list
A1 = A1[index_list]
A3 = A3[index_list]
A4 = A4[index_list]
values = values[index_list]
print(values)

#Values
mean_value = np.mean(values)
print('mean_value', mean_value)
sd_total = np.sqrt(np.sum((values - mean_value)**2) / (len(values) - 1)) #Berechnung der Standardabweichung SD
print('sd_total', sd_total)

(array([ 0,  5, 10, 11, 12, 13], dtype=int64),)
[0.28 0.8  0.46 0.39 0.28 1.1 ]
mean_value 0.5516666666666666
sd_total 0.32999494945630103


In [43]:
#Dataset A1
#For A
#calculate the mean value of A2 where A2 = A
#count A1 == 'A'

if(np.sum(A1 == 'A') <= 1):
    mean_A = 0;sd_A = 0
else:    
    mean_A = np.mean(values[A1 == 'A'])
    sd_A = np.sqrt(np.sum((values[A1 == 'A'] - mean_A)**2) / (len(values[A1 == 'A']) - 1)) #Berechnung der Standardabweichung SD
print('mean_A', mean_A)
print('sd_A', sd_A)
#For B
#calculate the mean value of A2 where A2 = B
if(np.sum(A1 == 'B') <= 1):
    mean_B = 0
    sd_B = 0
else:    
    mean_B = np.mean(values[A1 == 'B'])
    sd_B = np.sqrt(np.sum((values[A1 == 'B'] - mean_B)**2) / (len(values[A1 == 'B']) - 1)) #Berechnung der Standardabweichung SD
print('mean_B', mean_B)
print('sd_B', sd_B)
#For C
#calculate the mean value of A2 where A2 = C
if(np.sum(A1 == 'C') <= 1):
    mean_C = 0
    sd_C = 0
else:    
    mean_C = np.mean(values[A1 == 'C'])
    sd_C = np.sqrt(np.sum((values[A1 == 'C'] - mean_C)**2) / (len(values[A1 == 'C']) - 1)) #Berechnung der Standardabweichung SD
print('mean_C', mean_C)
print('sd_C', sd_C)

#standard deviation reduction
sdr_A1 = sd_total - np.sum ( len(values[A1 == 'A'])/len(values) * sd_A + len(values[A1 == 'B'])/len(values) * sd_B + len(values[A1 == 'C'])/len(values) * sd_C)
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A1', sdr_A1)

print('------------')

if(np.sum(A3 == 'T') <= 1):
    mean_T = 0;sd_T = 0
else:    
    mean_T = np.mean(values[A3 == 'T'])
    sd_T = np.sqrt(np.sum((values[A3 == 'T'] - mean_T)**2) / (len(values[A3 == 'T']) - 1)) #Berechnung der Standardabweichung SD
print('mean_R', mean_T)
print('sd_R', sd_T)
#For B
if(np.sum(A3 == 'S') <= 1):
    mean_S = 0
    sd_S = 0
else:    
    mean_S = np.mean(values[A3 == 'S'])
    sd_S = np.sqrt(np.sum((values[A3 == 'S'] - mean_S)**2) / (len(values[A3 == 'S']) - 1)) #Berechnung der Standardabweichung SD
print('mean_B', mean_S)
print('sd_B', sd_S)

if(np.sum(A3 == 'R') <= 1):
    mean_R = 0;sd_R = 0
else:    
    mean_R = np.mean(values[A3 == 'R'])
    sd_R = np.sqrt(np.sum((values[A3 == 'R'] - mean_R)**2) / (len(values[A3 == 'R']) - 1)) #Berechnung der Standardabweichung SD
print('mean_R', mean_R)
print('sd_R', sd_R)

#standard deviation reduction
sdr_A3 = sd_total - np.sum ( len(values[A3 == 'T'])/len(values) * sd_T + len(values[A3 == 'S'])/len(values) * sd_S + len(values[A3 == 'R'])/len(values) * sd_R)
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A3', sdr_A3)

print('------------')

if(np.sum(A4 == 'Z') <= 1):
    mean_Z = 0;sd_Z = 0
else:    
    mean_Z = np.mean(values[A4 == 'Z'])
    sd_Z = np.sqrt(np.sum((values[A4 == 'Z'] - mean_Z)**2) / (len(values[A4 == 'Z']) - 1)) #Berechnung der Standardabweichung SD
print('mean_Z', mean_Z)
print('sd_Z', sd_Z)
#For B
#calculate the mean value of A2 where A2 = B
if(np.sum(A4 == 'X') <= 1):
    mean_X = 0
    sd_X = 0
else:    
    mean_X = np.mean(values[A4 == 'X'])
    sd_X = np.sqrt(np.sum((values[A4 == 'X'] - mean_X)**2) / (len(values[A4 == 'X']) - 1)) #Berechnung der Standardabweichung SD
print('mean_X', mean_X)
print('sd_X', sd_X)

#standard deviation reduction
sdr_A4 = sd_total - np.sum ( len(values[A4 == 'Z'])/len(values) * sd_Z + len(values[A4 == 'X'])/len(values) * sd_X )
#sdr_A1 = sd_total - (len(values[A1 == 'A']) / len(values) * sd_A) - (len(values[A1 == 'B']) / len(values) * sd_B) - (len(values[A1 == 'C']) / len(values) * sd_C)
print('sdr_A4', sdr_A4)




mean_A 0.9500000000000001
sd_A 0.21213203435596428
mean_B 0.37
sd_B 0.12727922061357855
mean_C 0.335
sd_C 0.07778174593052022
sdr_A1 0.1909306158229467
------------
mean_R 0.335
sd_R 0.07778174593052022
mean_B 0.54
sd_B 0.36769552621700474
mean_R 0.78
sd_R 0.4525483399593905
sdr_A3 0.03065307875399592
------------
mean_Z 0.595
sd_Z 0.2899137802864845
mean_X 0.5533333333333333
sd_X 0.47342722073549315
sdr_A4 -0.0033565876736070255


b) Zeichnen Sie den Baum

In [ ]:
`

c) Bestimmen Sie den Mean-Squared-Error des Baumes
* auf den Trainingsdaten
* auf den folgenden Testdaten:

| A1 | A2 | A3 | A4 | Value |
| -- | -- | -- | -- | ----- |
| B  | J  | T  | Z  | 0.51  |
| C  | K  | R  | Y  | 1.90  |
| B  | J  | R  | X  | 0.90  |
| A  | J  | S  | Y  | 0.47  |
| A  | K  | T  | Z  | 0.54  |

Der Mean-Squared-Error wird folgendermaßen berechnet:

$$ MSE = \frac{1}{n} \sum_{j=1}^{n} ( y_j − p_j ) $$